crates/module-system/sov-modules-api/src/gas/traits.rs

```diff
@@ -944,6 +944,23 @@ impl<S: Spec> GetGasPrice for BasicGasMeter<S> {
     }
 }
 
+/// A subset of BasicGasMeter used to compute EVM gas limit
+pub struct BasicGasState<S: Spec> {
+    gas: S::Gas,
+    funds: Amount,
+    price: <<S as Spec>::Gas as Gas>::Price,
+}
+
+impl<S: Spec> BasicGasState<S> {
+    /// Computes gas limit as a min from gas and funds divided by gas price
+    pub fn gas_limit(&self) -> u64 {
+        min(
+            self.gas.as_ref()[0],
+            (self.funds.0 / self.price.as_ref()[0].0) as u64,
```

Comment:
This calulation is relevant only to the EVM. 
1. Let's make all the fields in `BasicGasState` public and make the calulation inside the evm module
2. `Amount` is `u128` so the division can overlow `u64`. Lets add the folloing logic
if self.funds = 0 & self.price.as_ref()[0].0 = 0 then gas =0
if self.funds != 0 & self.price.as_ref()[0].0 = 0 then gas = u64::MAX
if self.funds/self.price.as_ref()[0].0 overflows u64 then gas = u64::MAX